# Introduction to APIs (RESTful API Requests with Python)

Modern applications often need to communicate with other systems to exchange data.  
One of the most common ways this communication happens is through **APIs (Application Programming Interfaces)**.

An **API** is a set of rules and protocols that allows different software applications to communicate with each other. 

An API allows different software systems to exchange data.

## Examples
- Weather apps fetching live data  
- Apps retrieving user profiles  
- Websites processing payments  

## Key Idea
APIs act as a **bridge** between systems without exposing internal logic.

# What is a REST API?

**REST (Representational State Transfer)** is a standard way APIs communicate over the internet using HTTP.

## How it works
1. Client sends a request  
2. Server processes it  
3. Server returns data (usually JSON)  

## Analogy
You go to a restaurant (Client), a Waiter takes your order (Request) to the kitchen (Server) and brings back food (Data)

- Client → You  
- Request → Order  
- Server → Kitchen  
- Response → Food  

## Note
The server sends a **representation** (JSON), not raw database data.


# Key Components of REST APIs

## 1. Endpoint (URL)
The API address:
https://api.example.com/users

You must get this from the API documentation

---

## 2. HTTP Methods

| Method | Purpose |
|--------|--------|
| GET    | Retrieve data |
| POST   | Create data |
| PUT    | Update data |
| DELETE | Remove data |

---

## 3. Data Format (JSON)

Example:

{
  "name": "Alice",
  "age": 25
}


## Making API Requests in Python

Python provides the **requests library** to interact with REST APIs easily.

**First Step is to Install the requests library**

**Activate your Virtual Environment and then Install Requests**
## Install
pip install requests

**Then import requests**
## Import
import requests

**To use any API, you need to first check the API documentation**


# GET Request Example

A GET request retrieves data.

## Example (CoinGecko - simple endpoint)



In [ ]:
import requests



In [7]:
import requests

url = "https://api.coingecko.com/api/v3/ping" # gotten from the api documentation

headers = {"x-cg-demo-api-key": "CG-35uXYeJWDSqoeoMhdihMan2M"} # gotten from the api documentation

response = requests.get(url, headers=headers)

print(response.text)

{"gecko_says":"(V3) To the Moon!"}


# API Request Components (Important)

When making API requests, you often use:

---

## 1. URL (Endpoint)
The base address of the API.

Example:
https://api.coingecko.com/api/v3/ping

---

## 2. Params (Query Parameters)

Used to send **filters or options** in the URL.

Example URL:
https://api.com/data?user=1

Python:
params = {
    "user": 1
}

requests.get(url, params=params)

### When to use:
- Filtering data
- Searching
- Pagination

---

## 3. Headers

Headers send **extra information about the request**.

Example:
headers = {
    "Content-Type": "application/json",
    "Authorization": "Bearer API_KEY"
}

### Common Uses:
- Authentication (API keys)
- Telling server the data format
- Custom metadata

---

## 4. Body (Request Payload)

Used in POST/PUT requests to send data.

Example:
data = {
    "name": "Alice"
}

requests.post(url, json=data)

### Important:
- GET → no body
- POST/PUT → body required

---

## 5. Response

The server sends back:
- Data (JSON)
- Status code

Example:
response.json()
response.status_code

## POST Requests

A **POST request** sends data to the server to create or update resources.

Example use cases:

- Creating a user account
- Submitting form data
- Uploading information to a database

In [10]:
import requests

url = "https://jsonplaceholder.typicode.com/posts"

data = {
    "title": "Learning APIs",
    "body": "This is a POST request",
    "userId": 1
}

response = requests.post(url, json=data)

print(response.status_code)
print(response.json())


201
{'title': 'Learning APIs', 'body': 'This is a POST request', 'userId': 1, 'id': 101}


## HTTP Response Status Codes

When an API request is made, the server returns a **status code**.

Common status codes include:

| Code | Meaning |
|-----|------|
| 200 | Success |
| 201 | Resource created |
| 400 | Bad request |
| 404 | Resource not found |
| 500 | Server error |

## Handling Errors

API requests can fail due to:

- Network issues
- Invalid endpoints
- Server errors

Python allows error handling using if-else blocks.

In [11]:
# Error Handling

import requests

url = "https://api.coingecko.com/api/v3/ping"

response = requests.get(url)

if response.status_code == 200:
    print(response.json())
else:
    print("Request failed:", response.status_code)

{'gecko_says': '(V3) To the Moon!'}



---

# 📘 API Assignment Exercises

## 🟢 Exercise 1: Crypto Price Fetcher

**Goal:** Practice GET requests + params + JSON parsing + error handling

### Task:

* Use the CoinGecko API:

```
https://api.coingecko.com/api/v3/simple/price
```

* Fetch prices for:

  * Bitcoin
  * Ethereum
* Print Status code check and error message if request fails
* Print output like:

```
Bitcoin: $XXXXX
Ethereum: $XXXXX
```


## 🟡 Exercise 2: User Input + API Call

**Goal:** Make dynamic API requests

### Task:

* Ask the user to input a cryptocurrency name (e.g., bitcoin)
* Fetch its price in USD
* Print:

```
The price of <coin> is $XXXXX
```

### Hint:

Use `input()` and pass the value into `params`


## 🔴 Exercise 3: Mini API Script (Combine Everything)

**Goal:** Bring all concepts together

### Task:

Write a script that:

1. Asks user for a crypto name
2. Fetches price using API
3. Checks status code
4. Handles errors
5. Prints clean output

### Example:

```
Enter coin: bitcoin
Price of bitcoin is $XXXXX
```

If invalid coin:

```
Coin not found or API error
```

---



In [ ]:
import requests
# import json

url = "https://api.coingecko.com/api/v3/simple/price"

params = {
    "ids": "bitcoin,ethereum",
    "vs_currencies": "usd"
}

response = requests.get(url, params=params)

if response.status_code == 200:
    data = response.json()
    print("Bitcoin Price is", data["bitcoin"]["usd"])
    print("Ethereum Price is", data["ethereum"]["usd"])
else:
    print("Error fetching data")



Bitcoin Price is 74157
Ethereum Price is 2330.36
